added collect-to-arrow support

behrica opened the pull request:
refactored

more data types and refactored

more data types and refactored

make use of type hints

added docstring

clean up

adapted to latest upstream

coomit auto-changing file

behrica committed:
added collect tp arrow support

refactored

more data types and refactored

more data types and refactored

make use of type hints

added docstring

clean up

adapted to latest upstream

coomit auto-changing file

behrica committed:
added arrow dependencies

behrica committed:
fixed linting issues

anthony-khong commented:
Thank you for the PR! This looks great! Let me review the code 😄

anthony-khong commented on docker/project.clj:
```diff
@@ -33,16 +33,23 @@
                  [reply "0.4.4" :exclusions [javax.servlet/servlet-api]]
-                 [zero.one/fxl "0.0.5"]]
+                 [zero.one/fxl "0.0.5"]
+                 [org.apache.arrow/arrow-memory-netty "2.0.0"]
```
Perhaps this should be moved to provided dependencies? (all the Arrow deps I mean)

behrica replied:
done

anthony-khong commented on docker/project.clj:
```diff
@@ -33,16 +33,23 @@
    :dev {:dependencies [[enlive "1.1.6"]
-                        [midje "1.9.9"]]
+                        [midje "1.9.9"]
+                        [techascent/tech.ml.dataset "5.00-alpha-19"]
```
I believe this one should go to provided dependencies too.

behrica replied:
Having it in profile "dev" is not enough ?
This makes sure, is does not gets pulled in  using geni.

anthony-khong commented on docker/project.clj:
```diff
@@ -33,16 +33,23 @@
                    [lein-cloverage "1.2.1"]
                    [lein-midje "3.2.2"]
                    [lein-cljfmt "0.7.0"]]
+         :resource-paths ["test-resources"]
```
Why do we need this? Do we need to version control `test-resources/logback.xml`?

behrica replied:
Bringing in a "test" dependency to tech.ml.dataset made a lot of DEBUG logging appear.
Probably one of the dependencies of TMD has a concrete logger as dependency.
Not sure, if there is a better way to solve this.

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+(ns zero-one.geni.arrow
+  (:import [org.apache.arrow.vector VarCharVector
```
I believe you are getting unused imports from `clj-kondo`.

behrica replied:
fixed

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+    (when (nil? (value value-info))
+      (throw (Exception. "value cannot be null"))
+      ))
+  (case type
```
I think we I would prefer a nested map of functions, instead of the nested cases, but yea, this is a matter of taste I suppose. I'm happy with this if you feel strongly about it :)

behrica replied:
Does a "map of functions" work with  "calling java functions" ?
I remembered that we cannot do something like

(def m (hash-map f (.set))

The whole thing became as well "ugly", because I needed to have type-hints working.
They makes a huge difference in performance.


If you have an other proposal, how I can re-write it using:
- a map of functions
- type hints

On the other side, I am quite happy, that "missing" types now result in IllegalArgumentException.

behrica replied:
code is far better now.

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+
+;;; this method will be called for evey value transformed, so relection shoudkl be avoided
+(defn typed-action [action type value-info row-info ^String name ^RootAllocator allocator]
+  (when (= :set type)
```
Use `or` instead of double `when`s?

behrica replied:
is removed

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+    :boolean (case action
+               :set (.set ^BitVector (vector value-info) (idx1 value-info) 1 (if (value value-info) 1 0))
+               :make-vector (BitVector. name allocator)
+               :get (if (.isNullAt (row row-info) (idx2 row-info)) nil (.getBoolean (row row-info) (idx2 row-info))))
```
`when-not` instead of `(if p nil f)`?

behrica replied:
done

anthony-khong commented on src/clojure/zero_one/geni/core.clj:
```diff
@@ -83,7 +83,8 @@
    [zero-one.geni.spark]
-   [zero-one.geni.storage]))
+   [zero-one.geni.storage]
+   [zero-one.geni.arrow]))
```
Sort requires?

behrica replied:
done

anthony-khong commented on test/zero_one/geni/arrow_test.clj:
```diff
@@ -0,0 +1,90 @@
+              (g/read-parquet! "test/resources/withSqlDate.parquet")
+              (g/collect-to-arrow 10 temp-dir)))
+       (fact "dates are corect"
+             (let [with-date  (g/read-parquet! "test/resources/withSqlDate.parquet")
```
Can we use underscores for non-Java filenames?

behrica replied:
done

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+
+    ))
+
+(defn collect-to-arrow
```
I believe this is the only function we want in `core`. Should we make all the other functions private?

behrica replied:
done

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+  (:import [org.apache.arrow.vector VarCharVector
+            VectorSchemaRoot BigIntVector Float4Vector Float8Vector IntVector
+            BitVector ValueVector BaseFixedWidthVector TimeStampMilliVector]
+           [org.apache.arrow.vector.util Text]
```
Sort imports?

behrica replied:
Done

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+      ))
+  (case type
+    :string (case action
+              :set (.setSafe ^VarCharVector (vector value-info) (idx1 value-info) (Text. ^String (value value-info)))
```
I think you can make higher-level functions for `set`, `make-vector` and `get` to prevent any duplications. In particular, it's always `.set` or `setSafe` with `vector`, `idx1` and `value` called after.

behrica replied:
I will have a look.
I noticed the duplication, but when removing it, the type hints needed to stay. I am not super familiar with the type hints in clojure.

I was even thinking to eventually re-write the "typed-action" method in java. What do yo think ?
It is called for every single data element, so need to be fast.

anthony-khong replied:
I'd rather deal with Clojure with some duplications rather than Java to be honest! I may have a crack at it!

behrica replied:
I discovered, that "let" allows type hints, so the protocols could go away.
The code is far shorter know, and further duplication cannot be removed without giving up on type safety, which is needed in this place.

anthony-khong commented on src/clojure/zero_one/geni/arrow.clj:
```diff
@@ -0,0 +1,222 @@
+   #(hash-map
+     :type %1
+     :name %2)
+   (->> schema .fields (map #(keyword (.. % dataType typeName))))
```
Use `let`s to rename?

behrica replied:
I rewrote teh method

behrica committed:
fixed test file names

behrica committed:
made functions private

behrica committed:
refactored

behrica commented:
I addressed most of your points, thanks for comments.

I left function typed-action as-is.
I will have a look, on how to address duplication while preserving type hints.

behrica committed:
made arrow deps "provided"

behrica committed:
refactored type hints in typed-action

behrica committed:
removed not needed binding

behrica committed:
fixed whitespace

behrica committed:
reformated with cljfmt

anthony-khong committed:
Cosmetic changes

anthony-khong committed:
Another round of cosmetic changes

behrica committed:
typed hinted setNull

behrica committed:
use when-not

behrica committed:
implement and test emty dataframe

anthony-khong committed:
Merged from upstream + fixed formatting

anthony-khong committed:
Added Arrow dependencies to README + lein template

anthony-khong committed:
Made jsonista explicit dep

anthony-khong committed:
Added Arrow (+ other Spark deps) to deps.edn example

anthony-khong committed:
Updated leftover deps.edn for docker caching

behrica committed:
integrated Anthonys cosmetic changes

behrica commented:
I merged manually all your cosmetic changes.
So this branch should have them all.

behrica committed:
reformated

behrica commented:
Please have a look and lt me know, if fine.
I can the rebase it as well to squash all commits

behrica committed:
Merge branch 'collect_arrow' of https://github.com/anthony-khong/geni into anthony-khong-collect_arrow

behrica committed:
Merge branch 'anthony-khong-collect_arrow' into collect_arrow

anthony-khong commented:
Hi @behrica, I believe this is good to merge! Just one final thing to pass the CI jobs:

```
lein cljfmt fix src test doc
```

Then it should be good! Thank you for the awesome PR, this is a great addition to the library!

behrica committed:
reformated all

behrica commented:
Should I write some lines here: 

https://github.com/zero-one-group/geni/blame/6c540e4b19a0e888faecb8da4b5f8c52fc437c74/docs/cookbook/part_01_reading_and_writing_datasets.md#L122

to explain the different options Geni gives to "collect" data to the driver ?
(or in an other place)

We could mention as well, that there is now tight integration with TMD, as an other form to "collect" data to work with it further.

behrica committed:
added docu on collection of data

behrica committed:
fixed docu

behrica commented:
I added some docu , please have a look

anthony-khong commented:
Great! It's looking good! I'll merge as soon as the pipeline passes.

As for the docs, I suppose it's good to have. When we add TMD support in Geni, we should be able to do TMD-Spark interop from within Geni (i.e. no extra requires). When that happens, we just change the doc!

Again, thank you so much for this PR!!
